stl/inc/spanstream

```diff
@@ -0,0 +1,396 @@
+        : _Mysb(), _Mode(_Which), _Buf(_Span) {
+        if (_Mode & ios_base::out) {
+            if (_Mode & ios_base::ate) {
+                _Mysb::setp(_Buf.data(), _Buf.data() + _Buf.size(), _Buf.data() + _Buf.size());
+            } else {
+                _Mysb::setp(_Buf.data(), _Buf.data(), _Buf.data() + _Buf.size());
+            }
+        }
+
+        if (_Mode & ios_base::in) {
+            _Mysb::setg(_Buf.data(), _Buf.data(), _Buf.data() + _Buf.size());
+        }
+    }
+
+    basic_spanbuf(const basic_spanbuf&) = delete;
+    basic_spanbuf(basic_spanbuf&& _Right)
+        : _Mysb(_STD move(_Right)), _Mode(_STD move(_Right._Mode)), _Buf(_STD move(_Right._Buf)) {
+        _Right.setp(nullptr, nullptr, nullptr);
+        _Right.setg(nullptr, nullptr, nullptr);
+    }
+
+    // [spanbuf.assign], assignment and swap
+    basic_spanbuf& operator=(const basic_spanbuf&) = delete;
+
+    basic_spanbuf& operator=(basic_spanbuf&& _Right) {
+        basic_spanbuf tmp{move(_Right)};
+        this->swap(tmp);
+        return *this;
+    }
+
+    void swap(basic_spanbuf& _Right) {
+        _Mysb::swap(_Right);
+        _STD swap(_Mode, _Right._Mode);
+        _STD swap(_Buf, _Right._Buf);
+    }
+
+    // [spanbuf.members], member functions
+    _STD span<_Elem> span() const noexcept {
+        if (_Mode & ios_base::out) {
+            return _STD span<_Elem>(_Mysb::pbase(), _Mysb::pptr());
+        }
+
+        return _Buf;
+    }
+
+    void span(_STD span<_Elem> _Span) noexcept {
+        _Buf = _Span;
+        if (_Mode & ios_base::out) {
+            if (_Mode & ios_base::ate) {
+                _Mysb::setp(_Buf.data(), _Buf.data() + _Buf.size(), _Buf.data() + _Buf.size());
+            } else {
+                _Mysb::setp(_Buf.data(), _Buf.data(), _Buf.data() + _Buf.size());
+            }
+        }
+
+        if (_Mode & ios_base::in) {
+            _Mysb::setg(_Buf.data(), _Buf.data(), _Buf.data() + _Buf.size());
+        }
+    }
+
+protected:
+    // [spanbuf.virtuals], overridden virtual functions
+    pos_type __CLR_OR_THIS_CALL seekoff(
+        off_type _Off, ios_base::seekdir _Way, ios_base::openmode _Which = ios_base::in | ios_base::out) override {
+        const bool _Sequence_in  = _Which & ios_base::in;
+        const bool _Sequence_out = _Which & ios_base::out;
+        switch (_Way) {
+        case ios_base::beg:
+            {
+                // [spanbuf.virtuals]/4.1 baseoff = 0
+                if (static_cast<size_t>(_Off) > _Buf.size()) { // negative wraparound to positive to save a compare
+                    return pos_type(off_type(-1)); // [spanbuf.virtuals]/5 report failure
+                }
+
+                break;
+            }
+        case ios_base::end:
+            {
+                // [spanbuf.virtuals]/4.3 baseoff =
+                const auto _Baseoff = _Mode & ios_base::out && !(_Mode & ios_base::in)

```

Comment:
This is mixing bitwise and logical operators, so we conventionally use parentheses to disambiguate the (notoriously confusing for humans) precedence: `(_Mode & ios_base::out)`